library/data/list.lean

```diff
@@ -41,6 +41,25 @@ theorem length_dropn
           = length l - i             : length_dropn i l
       ... = succ (length l) - succ i : nat.sub_eq_succ_sub_succ (length l) i
 
+definition has_decidable_eq [h : decidable_eq α]
+: ∀ (x y : list α), decidable (x = y)
+| nil nil := is_true rfl
+| nil (cons b s) := is_false (λ q, list.no_confusion q)
+| (cons a r) nil := is_false (λ q, list.no_confusion q)
+| (cons a r) (cons b s) :=
+  match h a b with
+  | (is_true h₁) :=
+    match has_decidable_eq r s with
+    | (is_true  h₂) :=
+       is_true (calc a :: r = b :: r : congr_arg (λc, c :: r) h₁
+                        ... = b :: s : congr_arg (λt, b :: t) h₂)
+    | (is_false h₂) :=
+      is_false (λ q, list.no_confusion q (λ heq teq, h₂ teq))
+    end
+  | (is_false h₁) :=
+     is_false (λ q, list.no_confusion q (λ heq teq, h₁ heq))
+  end
+
```

Comment:
Thanks, I pushed a commit that removes the redundant decidable instances; I had these in a fork that had gotten stale.